File: 2. Add Two Numbers (Medium)/solution.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def addTwoNumbers(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    result = ListNode()
    current = result
    carry = 0
    
    while l1 or l2 or carry:
        sum_val = carry
        if l1:
            sum_val += l1.val
            l1 = l1.next
        if l2:
            sum_val += l2.val
            l2 = l2.next
        
        carry, val = divmod(sum_val, 10)
        current.next = ListNode(val)
        current = current.next
    
    return result.next



l1 = ListNode(2, ListNode(4, ListNode(3)))
l2 = ListNode(5, ListNode(6, ListNode(4)))
result = addTwoNumbers(l1, l2)
while result:
    print(result.val, end=" ")
    result = result.next

File: 2. Add Two Numbers (Medium)/solution.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	current := dummy
	carry := 0

	for l1 != nil || l2 != nil || carry != 0 {
		sumVal := carry
		if l1 != nil {
			sumVal += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sumVal += l2.Val
			l2 = l2.Next
		}

		carry = sumVal / 10
		current.Next = &ListNode{Val: sumVal % 10}
		current = current.Next
	}

	return dummy.Next
}

func main() {
	l1 := &ListNode{Val: 2, Next: &ListNode{Val: 4, Next: &ListNode{Val: 3}}}
	l2 := &ListNode{Val: 5, Next: &ListNode{Val: 6, Next: &ListNode{Val: 4}}}
	result := addTwoNumbers(l1, l2)
	for result != nil {
		fmt.Print(result.Val, " ")
		result = result.Next
	}
}
